File: examples/sample_app/ecommerce.js
/**
 * ecommerce.js – Sample e-commerce application (JavaScript / ES2020)
 *
 * Mirrors ecommerce.py with the same three distinct application flows:
 *   1. registerUser   – create account, verify email
 *   2. checkout       – build cart, apply coupon, process payment, create order
 *   3. trackOrder     – look up order, update status, send notification
 *
 * Used by FlowDelta to demonstrate AST analysis and flow identification
 * on JavaScript source code.
 */

'use strict';

const crypto = require('crypto');

// ---------------------------------------------------------------------------
// Domain models
// ---------------------------------------------------------------------------

class User {
  constructor(email, passwordHash) {
    this.userId = crypto.randomBytes(4).toString('hex');
    this.email = email;
    this.passwordHash = passwordHash;
    this.verified = false;
    this.createdAt = new Date().toISOString();
  }
}

class Product {
  constructor(productId, name, price, stock) {
    this.productId = productId;
    this.name = name;
    this.price = price;
    this.stock = stock;
  }
}

class CartItem {
  constructor(product, quantity) {
    this.product = product;
    this.quantity = quantity;
  }

  get subtotal() {
    return Math.round(this.product.price * this.quantity * 100) / 100;
  }
}

class Cart {
  constructor(userId) {
    this.userId = userId;
    this.items = [];
    this.couponCode = null;
    this.discount = 0;
  }

  get total() {
    const subtotal = this.items.reduce((sum, i) => sum + i.subtotal, 0);
    return Math.round((subtotal - this.discount) * 100) / 100;
  }
}

class Order {
  constructor(orderId, userId, items, total) {
    this.orderId = orderId;
    this.userId = userId;
    this.items = items;
    this.total = total;
    this.status = 'pending';
    this.paymentRef = null;
    this.updatedAt = new Date().toISOString();
  }
}

// ---------------------------------------------------------------------------
// In-memory "database"
// ---------------------------------------------------------------------------

const _users = {};
const _products = {
  p001: new Product('p001', 'Laptop',     999.99, 5),
  p002: new Product('p002', 'Headphones',  49.99, 20),
  p003: new Product('p003', 'Mouse',       19.99, 50),
};
const _orders = {};
const _coupons = { SAVE10: 10, SUMMER20: 20 };

// ---------------------------------------------------------------------------
// Flow 1 – User Registration
// ---------------------------------------------------------------------------

function hashPassword(password) {
  return crypto.createHash('sha256').update(password).digest('hex');
}

function createAccount(email, password) {
  if (_users[email]) {
    throw new Error(`Email already registered: ${email}`);
  }
  const user = new User(email, hashPassword(password));
  _users[email] = user;
  return user;
}

function sendVerificationEmail(user) {
  // Simulate sending – returns a 6-char token
  const token = crypto.createHash('md5').update(user.email).digest('hex').slice(0, 6).toUpperCase();
  return token;
}

function verifyEmail(user, token) {
  const expected = crypto.createHash('md5').update(user.email).digest('hex').slice(0, 6).toUpperCase();
  if (token === expected) {
    user.verified = true;
    _users[user.email] = user;
    return true;
  }
  return false;
}

function registerUser(email, password) {
  const user = createAccount(email, password);
  const token = sendVerificationEmail(user);
  verifyEmail(user, token);
  return user;
}

// ---------------------------------------------------------------------------
// Flow 2 – Product Checkout
// ---------------------------------------------------------------------------

function buildCart(userId, productQuantities) {
  const cart = new Cart(userId);
  for (const [pid, qty] of Object.entries(productQuantities)) {
    const product = _products[pid];
    if (!product) throw new Error(`Unknown product: ${pid}`);
    if (product.stock < qty) throw new Error(`Insufficient stock for ${pid}`);
    cart.items.push(new CartItem(product, qty));
  }
  return cart;
}

function applyCoupon(cart, couponCode) {
  const discount = _coupons[couponCode];
  if (discount === undefined) throw new Error(`Invalid coupon: ${couponCode}`);
  cart.couponCode = couponCode;
  cart.discount = discount;
  return cart;
}

function processPayment(cart, cardLast4) {
  if (cart.total <= 0) throw new Error('Cart total must be positive');
  const paymentRef = 'PAY-' + crypto.randomBytes(4).toString('hex').toUpperCase();
  // Deduct stock
  for (const item of cart.items) {
    _products[item.product.productId].stock -= item.quantity;
  }
  return paymentRef;
}

function createOrder(cart, paymentRef) {
  const orderId = 'ORD-' + crypto.randomBytes(3).toString('hex').toUpperCase();
  const order = new Order(orderId, cart.userId, [...cart.items], cart.total);
  order.paymentRef = paymentRef;
  order.status = 'confirmed';
  _orders[orderId] = order;
  return order;
}

function checkout(userId, productQuantities, coupon = null) {
  let cart = buildCart(userId, productQuantities);
  if (coupon) {
    cart = applyCoupon(cart, coupon);
  }
  const paymentRef = processPayment(cart, '4242');
  const order = createOrder(cart, paymentRef);
  return order;
}

// ---------------------------------------------------------------------------
// Flow 3 – Order Tracking
// ---------------------------------------------------------------------------

const VALID_TRANSITIONS = {
  confirmed: ['shipped'],
  shipped:   ['delivered'],
  delivered: ['closed'],
};

function lookupOrder(orderId) {
  const order = _orders[orderId];
  if (!order) throw new Error(`Order not found: ${orderId}`);
  return order;
}

function updateOrderStatus(order, newStatus) {
  const allowed = VALID_TRANSITIONS[order.status] || [];
  if (!allowed.includes(newStatus)) {
    throw new Error(
      `Invalid transition: ${order.status} → ${newStatus}. Allowed: ${allowed}`
    );
  }
  order.status = newStatus;
  order.updatedAt = new Date().toISOString();
  _orders[order.orderId] = order;
  return order;
}

function sendStatusNotification(order) {
  return {
    recipient: order.userId,
    message:   `Your order ${order.orderId} is now ${order.status}.`,
    sent:      true,
  };
}

function trackOrder(orderId, newStatus) {
  const order = lookupOrder(orderId);
  updateOrderStatus(order, newStatus);
  const notification = sendStatusNotification(order);
  return { order, notification };
}

// ---------------------------------------------------------------------------
// Module exports
// ---------------------------------------------------------------------------

module.exports = {
  // Models
  User, Product, CartItem, Cart, Order,
  // Flow 1
  hashPassword, createAccount, sendVerificationEmail, verifyEmail, registerUser,
  // Flow 2
  buildCart, applyCoupon, processPayment, createOrder, checkout,
  // Flow 3
  lookupOrder, updateOrderStatus, sendStatusNotification, trackOrder,
};

File: examples/sample_app/ecommerce.py
"""
Sample e-commerce application used to demonstrate FlowDelta.

Contains three distinct flows:
  1. user-registration   – create account, verify email
  2. product-checkout    – add to cart, apply coupon, process payment
  3. order-tracking      – look up order, update status, send notification
"""

from __future__ import annotations

import hashlib
import uuid
from dataclasses import dataclass, field
from datetime import datetime
from typing import Dict, List, Optional


# ---------------------------------------------------------------------------
# Domain models
# ---------------------------------------------------------------------------

@dataclass
class User:
    email: str
    password_hash: str
    user_id: str = field(default_factory=lambda: str(uuid.uuid4())[:8])
    verified: bool = False
    created_at: str = field(default_factory=lambda: datetime.utcnow().isoformat())


@dataclass
class Product:
    product_id: str
    name: str
    price: float
    stock: int


@dataclass
class CartItem:
    product: Product
    quantity: int

    @property
    def subtotal(self) -> float:
        return round(self.product.price * self.quantity, 2)


@dataclass
class Cart:
    user_id: str
    items: List[CartItem] = field(default_factory=list)
    coupon_code: Optional[str] = None
    discount: float = 0.0

    @property
    def total(self) -> float:
        subtotal = sum(i.subtotal for i in self.items)
        return round(subtotal - self.discount, 2)


@dataclass
class Order:
    order_id: str
    user_id: str
    items: List[CartItem]
    total: float
    status: str = "pending"
    payment_ref: Optional[str] = None
    updated_at: str = field(default_factory=lambda: datetime.utcnow().isoformat())


# ---------------------------------------------------------------------------
# In-memory "database"
# ---------------------------------------------------------------------------

_users: Dict[str, User] = {}
_products: Dict[str, Product] = {
    "p001": Product("p001", "Laptop",   999.99, 5),
    "p002": Product("p002", "Headphones", 49.99, 20),
    "p003": Product("p003", "Mouse",     19.99, 50),
}
_orders: Dict[str, Order] = {}
_coupons = {"SAVE10": 10.0, "SUMMER20": 20.0}


# ---------------------------------------------------------------------------
# Flow 1 – User Registration
# ---------------------------------------------------------------------------

def hash_password(password: str) -> str:
    return hashlib.sha256(password.encode()).hexdigest()


def create_account(email: str, password: str) -> User:
    if email in _users:
        raise ValueError(f"Email already registered: {email}")
    user = User(email=email, password_hash=hash_password(password))
    _users[email] = user
    return user


def send_verification_email(user: User) -> str:
    """Simulate sending verification; returns a token."""
    token = hashlib.md5(user.email.encode()).hexdigest()[:6].upper()
    return token


def verify_email(user: User, token: str) -> bool:
    expected = hashlib.md5(user.email.encode()).hexdigest()[:6].upper()
    if token == expected:
        user.verified = True
        _users[user.email] = user
        return True
    return False


def register_user(email: str, password: str) -> User:
    """Entry point – Flow 1: user-registration."""
    user = create_account(email, password)
    token = send_verification_email(user)
    verify_email(user, token)
    return user


# ---------------------------------------------------------------------------
# Flow 2 – Product Checkout
# ---------------------------------------------------------------------------

def build_cart(user_id: str, product_quantities: Dict[str, int]) -> Cart:
    cart = Cart(user_id=user_id)
    for pid, qty in product_quantities.items():
        product = _products.get(pid)
        if not product:
            raise ValueError(f"Unknown product: {pid}")
        if product.stock < qty:
            raise ValueError(f"Insufficient stock for {pid}")
        cart.items.append(CartItem(product=product, quantity=qty))
    return cart


def apply_coupon(cart: Cart, coupon_code: str) -> Cart:
    discount = _coupons.get(coupon_code, 0.0)
    if discount == 0.0:
        raise ValueError(f"Invalid coupon: {coupon_code}")
    cart.coupon_code = coupon_code
    cart.discount = discount
    return cart


def process_payment(cart: Cart, card_last4: str) -> str:
    """Simulate payment processing; returns a payment reference."""
    if cart.total <= 0:
        raise ValueError("Cart total must be positive")
    payment_ref = f"PAY-{uuid.uuid4().hex[:8].upper()}"
    # Deduct stock
    for item in cart.items:
        _products[item.product.product_id].stock -= item.quantity
    return payment_ref


def create_order(cart: Cart, payment_ref: str) -> Order:
    order = Order(
        order_id=f"ORD-{uuid.uuid4().hex[:6].upper()}",
        user_id=cart.user_id,
        items=list(cart.items),
        total=cart.total,
        payment_ref=payment_ref,
        status="confirmed",
    )
    _orders[order.order_id] = order
    return order


def checkout(user_id: str, product_quantities: Dict[str, int], coupon: Optional[str] = None) -> Order:
    """Entry point – Flow 2: product-checkout."""
    cart = build_cart(user_id, product_quantities)
    if coupon:
        cart = apply_coupon(cart, coupon)
    payment_ref = process_payment(cart, card_last4="4242")
    order = create_order(cart, payment_ref)
    return order


# ---------------------------------------------------------------------------
# Flow 3 – Order Tracking
# ---------------------------------------------------------------------------

def lookup_order(order_id: str) -> Order:
    order = _orders.get(order_id)
    if not order:
        raise ValueError(f"Order not found: {order_id}")
    return order


def update_order_status(order: Order, new_status: str) -> Order:
    valid_transitions = {
        "confirmed": ["shipped"],
        "shipped": ["delivered"],
        "delivered": ["closed"],
    }
    allowed = valid_transitions.get(order.status, [])
    if new_status not in allowed:
        raise ValueError(
            f"Invalid transition: {order.status} → {new_status}. Allowed: {allowed}"
        )
    order.status = new_status
    order.updated_at = datetime.utcnow().isoformat()
    _orders[order.order_id] = order
    return order


def send_status_notification(order: Order) -> dict:
    """Simulate sending an email/SMS notification."""
    return {
        "recipient": order.user_id,
        "message": f"Your order {order.order_id} is now {order.status}.",
        "sent": True,
    }


def track_order(order_id: str, new_status: str) -> dict:
    """Entry point – Flow 3: order-tracking."""
    order = lookup_order(order_id)
    order = update_order_status(order, new_status)
    notification = send_status_notification(order)
    return {"order": order, "notification": notification}
